recipes/upgrade.rb

```diff
@@ -24,31 +24,37 @@
   raise 'Failed to upgrade'
 end
 
-service 'splunk_stop' do
-  service_name 'splunk'
-  supports status: true
-  provider Chef::Provider::Service::Init
-  action :stop
+# ensure that the splunk service resource is available without cloning
+# the resource (CHEF-3694). this is so the later notification works,
+# especially when using chefspec to run this cookbook's specs.
+begin
+  resources('service[splunk]')
+rescue Chef::Exceptions::ResourceNotFound
+  service 'splunk' do
+    supports status: true, restart: true
+    provider splunk_service_provider
+    action :nothing
+  end
 end
```

Comment:
This is really, really odd to read in recipe code.

This likely indicates that this should be moved into a custom resource -- you can create a splunk_update resource similarly to build_essentials which takes no argument.

Reply:
At any rate this should use either find_resource with a block or edit_resource with a block

Reply:
reading a bit closer I'd say this definitely needs to be moved into the custom resource and the notifications hanging off of the upgrade action in the recipe code need to be removed and all that needs to be handled from within the custom resource.